README.md
Eve-Embedded
===========

[![Build Status](https://travis-ci.org/luiscoms/eve-embedded.svg?branch=master)](https://travis-ci.org/luiscoms/eve-embedded)
[![Requirements Status](https://requires.io/github/luiscoms/eve-embedded/requirements.svg?branch=master)](https://requires.io/github/luiscoms/eve-embedded/requirements/?branch=master)
[![Coverage Status](https://coveralls.io/repos/github/luiscoms/eve-embedded/badge.svg?branch=master)](https://coveralls.io/github/luiscoms/eve-embedded?branch=master)

This project servers as an example of a Python package with some boilerplate
code already in place.


Usage
----

In the schema set the url

```python
schema = {
    "firstname": {
        "type": "string",
        "minlength": 1,
        "maxlength": 10,
    },
    "country": {
        "type": "string",
        "schema": {
            "type": "string",
            "data_relation": {
                "api": "http://api.example.com/country",
                "embeddable": True
            }
        }
    }
}
```

Then install the module

```python
from eve_embedded import embedded

app = Eve()
embedded.install(app)
```


Special field types
----

When you are using specia field types, you can pass it as argumen like:

```python

skill_schema = {
    "title": {
        "type": "string",
    },
    "level": {
        "type": "string",
        "schema": {
            "type": "string",
            "data_relation": {
                "api": "http://api.example.com/levels",
                "embeddable": True
            }
        }
    }
}

skill_type_schema = {
    "type": "skill",
    "schema": skill_schema
}

schema = {
    "firstname": {
        "type": "string",
        "minlength": 1,
        "maxlength": 10,
    },
    "country": {
        "type": "string",
        "schema": {
            "type": "string",
            "data_relation": {
                "api": "http://api.example.com/country",
                "embeddable": True
            }
        }
    },
    "skills": {
        "type":  "list",
        "schema": {
            "type":  "skill",
        }
    }
}

app = Eve()
embedded.install(app, dict(area=skill_type_schema))
```

Running tests with `tox`
----

You will need an instance of `mongo` running locally

Install `tox`
```
$ pip install tox
```

Run tests

```
tox
```


CHANGELOG.md
Eve-Embedded
===========

0.1.0
-----

* Add `special_field_types` argument when install

0.0.3
-----

* Fixing embedded on areas automaticas

0.0.2
-----

* Removing quote from no-cache header

0.0.1
-----

* Initial release


tests/__init__.py
import json
import os
import unittest

import eve
from bson import ObjectId
from flask_pymongo import MongoClient

import eve_embedded.embedded
from tests.test_settings import \
    MONGO_HOST, MONGO_PORT, \
    MONGO_USERNAME, MONGO_PASSWORD, MONGO_DBNAME


class TestBase(unittest.TestCase):

    test_data = {
        "skills": [
            {
                '_id': ObjectId('570c24f8de9f0c5e6c7a2e71'),
                'name': "Programming"
            },
            {
                '_id': ObjectId('570c24fbde9f0c5e6c7a2e9f'),
                'name': "Reading"
            },
        ],
        "people": [
            {
                '_id': ObjectId('54f112defba522406c9cc209'),
                'name': "John",
                "skills": [ObjectId('570c24f8de9f0c5e6c7a2e71'), ObjectId('570c24fbde9f0c5e6c7a2e9f')]
            },
            {
                '_id': ObjectId('585a8be8f0983235bf0f95ed'),
                'name': "Peter",
                "rest_skills": ['570c24f8de9f0c5e6c7a2e71', '570c24fbde9f0c5e6c7a2e9f']
            }
        ]
    }

    def setUp(self, settings=None):
        self.this_directory = os.path.dirname(os.path.realpath(__file__))
        if settings is None:
            settings = os.path.join(self.this_directory, 'test_settings.py')

        self.setupDB()

        self.settings = settings
        self.app = eve.Eve(settings=self.settings)
        eve_embedded.embedded.install(self.app)

        self.test_client = self.app.test_client()
        self.domain = self.app.config['DOMAIN']

    def tearDown(self):
        del self.app
        self.dropDB()

    def setupDB(self):
        self.connection = MongoClient(MONGO_HOST, MONGO_PORT)
        self.connection.drop_database(MONGO_DBNAME)
        if MONGO_USERNAME:
            self.connection[MONGO_DBNAME].add_user(MONGO_USERNAME,
                                                   MONGO_PASSWORD)

        # seed
        self.connection[MONGO_DBNAME] \
            .skills \
            .insert_many(self.test_data.get("skills"))
        self.connection[MONGO_DBNAME] \
            .people \
            .insert_many(self.test_data.get("people"))

    def dropDB(self):
        self.connection = MongoClient(MONGO_HOST, MONGO_PORT)
        self.connection.drop_database(MONGO_DBNAME)
        self.connection.close()

    def parse_response(self, r):
        try:
            v = json.loads(r.get_data().decode('utf-8'))
        except ValueError:
            v = None
        return v


tests/test_app.py
from tests import TestBase

t = TestBase()
t.setUp()

application = t.app

if __name__ == '__main__':
    t.app.run(host='0.0.0.0', port=8080, debug=True)
    t.tearDown()


tests/test_example.py
import json
from pprint import pprint

import httpretty

from tests import TestBase  # noqa: E402


class TestEveEmbedded(TestBase):
    def test_api(self):
        self.app.debug = True
        r = self.test_client.get('/')
        self.assertEqual(r.status_code, 200)

    def test_resource(self):
        self.app.debug = True
        r = self.test_client.get('/people')
        self.assertEqual(r.status_code, 200)

    def test_json(self):
        self.app.debug = True
        r = self.test_client.get('/people')
        s = r.get_data().decode('utf-8')  # raises UnicodeError
        j = json.loads(s)  # raises ValueError
        # pprint(j)

    # def notest_doc_is_dict(self):
    def test_embedded_relations(self):
        doc = self.test_client\
            .get('/people?embedded={"relations.relation":1}')\
            .get_data()\
            .decode('utf-8')

        j = json.loads(doc)
        print("--ls--"*10)
        pprint(j)
        self.assertIn('_items', doc)
        # self.assertIsInstance(doc, dict)

    @httpretty.activate
    def test_embedded_rest_relations(self):
        httpretty.disable()
        print("!!ls!!"*10)
        expected = self.test_client\
                       .get('/people')\
                       .get_data()\
                       .decode('utf-8')
        expected_json = json.loads(expected)
        pprint(expected_json)

        httpretty.enable()
        import re
        httpretty.register_uri(
            httpretty.GET,
            re.compile("http://myapi.com/(.*)"),
            body="{}",
        )
        httpretty.register_uri(httpretty.GET,
                               'http://myapi.com/people?embedded={"rest_relations.relation":1}',
                               content_type="application/json",
                               body=expected)
        print("register http://myapi.com/people/54f112defba522406c9cc209", type(json.dumps(expected_json.get('_items')[0])))
        httpretty.register_uri(httpretty.GET,
                               'http://myapi.com/people/54f112defba522406c9cc209',
                               content_type="application/json",
                               body=json.dumps(expected_json.get('_items')[0]))
        httpretty.register_uri(httpretty.GET,
                               'http://myapi.com/people/585a8be8f0983235bf0f95ed',
                               content_type="application/json",
                               body=json.dumps(expected_json.get('_items')[1]))

        doc = self.test_client \
            .get('/people?embedded={"rest_relations.relation":1}') \
            .get_data() \
            .decode('utf-8')

        print(">>> last_request")
        pprint(httpretty.last_request().body)
        print("<<< last_request")

        response_json = json.loads(doc)
        print("--mocked--" * 10)
        pprint(response_json)

        self.assertIn('_items', doc)
        self.assertEqual(expected_json, response_json)
        # self.assertIsInstance(doc, dict)

        # httpretty.disable()
        # httpretty.reset()

    def notest_info(self):
        doc = self.swagger_doc

        self.assertIn('info', doc)
        self.assertIn('title', doc['info'])
        self.assertTrue(isinstance(doc['info']['title'], u''.__class__))
        self.assertIn('version', doc['info'])
        self.assertTrue(isinstance(doc['info']['version'], u''.__class__))

    def notest_paths(self):
        doc = self.swagger_doc
        url = self.domain['people']['url']
        item_title = self.domain['people']['item_title']

        self.assertIn('paths', doc)
        self.assertIsInstance(doc['paths'], dict)
        self.assertIn('/'+url, doc['paths'])
        self.assertIn('/%s/{%sId}' % (url, item_title.lower()), doc['paths'])

    def notest_definitions(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']

        self.assertIn('definitions', doc)
        self.assertIsInstance(doc['definitions'], dict)
        self.assertIn(item_title, doc['definitions'])
        self.assertIn('properties', doc['definitions'][item_title])
        self.assertEqual(
            set(doc['definitions'][item_title]['properties'].keys()),
            set(['name', 'job', '_id', 'relations']))

    def notest_parameters_people(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']
        lookup_field = self.domain['people']['item_lookup_field']

        self.assertIn('parameters', doc)
        self.assertIn(item_title+'_'+lookup_field, doc['parameters'])

    def notest_resource_description(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']

        self.assertEqual('the people resource',
                         doc['definitions'][item_title]['description'])

    def notest_field_description(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']
        props = doc['definitions'][item_title]['properties']

        self.assertEqual('the last name of the person',
                         props['name']['description'])

    def notest_disabled_resource(self):
        doc = self.swagger_doc
        url = self.domain['disabled_resource']['url']
        item_title = self.domain['disabled_resource']['item_title']
        lookup_field = self.domain['disabled_resource']['item_lookup_field']

        self.assertNotIn('/'+url, doc['paths'])
        self.assertNotIn('/%s/{%sId}' % (url, item_title.lower()),
                         doc['paths'])
        self.assertNotIn(item_title, doc['definitions'])
        self.assertNotIn(item_title+'_'+lookup_field, doc['parameters'])

    def notest_data_relation_source_field(self):
        doc = self.swagger_doc

        source_field = self.domain['people']['item_title']+'_job'
        self.assertIn(source_field, doc['definitions'])

    def notest_reference_to_data_relation_source_field(self):
        doc = self.swagger_doc
        people_it = self.domain['people']['item_title']
        people_props = doc['definitions'][people_it]['properties']
        dr_1_it = self.domain['dr_resource_1']['item_title']
        dr_1_props = doc['definitions'][dr_1_it]['properties']
        key = people_it + '_job'

        self.assertIn('$ref', people_props['job'])
        self.assertEqual('#/definitions/%s' % key,
                         people_props['job']['$ref'])
        self.assertIn(key, doc['definitions'])

        self.assertIn('$ref', dr_1_props['copied_field_with_description'])
        self.assertEqual('#/definitions/%s' % key,
                         dr_1_props['copied_field_with_description']['$ref'])

        key = people_it + '__id'
        people_rels_props = people_props['relations']['items']['properties']
        self.assertIn('$ref', people_rels_props['relation'])
        self.assertEqual('#/definitions/%s' % key,
                         people_rels_props['relation']['$ref'])
        self.assertIn(key, doc['definitions'])

    def notest_data_relation_extended_description(self):
        doc = self.swagger_doc
        item_title = self.domain['dr_resource_1']['item_title']
        lookup_field = self.domain['dr_resource_1']['item_lookup_field']
        par = doc['parameters'][item_title+'_'+lookup_field]
        people_it = self.domain['people']['item_title']

        self.assertIn('description', par)
        self.assertEqual(
            'foobar copied_field (links to {0}_job)'.format(people_it),
            par['description'])

    def notest_data_relation_copied_description(self):
        doc = self.swagger_doc
        item_title = self.domain['dr_resource_2']['item_title']
        lookup_field = self.domain['dr_resource_2']['item_lookup_field']
        par = doc['parameters'][item_title+'_'+lookup_field]
        people_it = self.domain['people']['item_title']

        self.assertIn('description', par)
        self.assertEqual(
            'the job of the person (links to {0}_job)'.format(people_it),
            par['description'])

    def notest_header_parameters(self):
        doc = self.swagger_doc
        url = self.domain['people']['url']
        item_title = self.domain['people']['item_title']
        url = '/%s/{%sId}' % (url, item_title.lower())

        header_parameters = []
        # assume that header parameters are equal for PUT, PATCH, and DELETE
        for method in ['put', 'patch', 'delete']:
            for p in doc['paths'][url][method]['parameters']:
                if 'in' not in p:
                    continue
                if p['in'] == 'header':
                    if method in ['patch', 'delete']:
                        # already added in 'put'
                        self.assertIn(p, header_parameters)
                    else:
                        header_parameters += [p]

        self.assertTrue(len(header_parameters) == 1)
        h = header_parameters[0]
        self.assertIn('name', h)
        self.assertEqual(h['name'], 'If-Match')

    def notest_cors_without_origin(self):
        self.app.config['X_DOMAINS'] = ['http://example.com']
        self.app.config['X_HEADERS'] = ['Origin', 'X-Requested-With',
                                        'Content-Type', 'Accept']
        self.app.config['X_MAX_AGE'] = 2000
        r = self.test_client.get('/api-docs')

        self.assertEqual(r.status_code, 200)
        self.assertNotIn('Access-Control-Allow-Origin', r.headers)
        self.assertNotIn('Access-Control-Allow-Headers', r.headers)
        self.assertNotIn('Access-Control-Allow-Methods', r.headers)
        self.assertNotIn('Access-Control-Max-Age', r.headers)
        self.assertNotIn('Access-Control-Expose-Headers', r.headers)

    def notest_cors_with_origin(self):
        self.app.config['X_DOMAINS'] = 'http://example.com'
        self.app.config['X_HEADERS'] = ['Origin', 'X-Requested-With',
                                        'Content-Type', 'Accept']
        self.app.config['X_MAX_AGE'] = 2000
        r = self.test_client.get('/api-docs',
                                 headers={'Origin': 'http://example.com'})

        self.assertEqual(r.status_code, 200)
        self.assertIn('Access-Control-Allow-Origin', r.headers)
        self.assertEqual(r.headers['Access-Control-Allow-Origin'],
                         'http://example.com')
        self.assertIn('Access-Control-Allow-Headers', r.headers)
        self.assertEqual(
            set(r.headers['Access-Control-Allow-Headers'].split(', ')),
            set(['Origin', 'X-Requested-With', 'Content-Type', 'Accept']))
        self.assertIn('Access-Control-Allow-Methods', r.headers)
        self.assertEqual(
            set(r.headers['Access-Control-Allow-Methods'].split(', ')),
            set(['HEAD', 'OPTIONS', 'GET']))
        self.assertIn('Access-Control-Max-Age', r.headers)
        self.assertEqual(r.headers['Access-Control-Max-Age'],
                         '2000')

    def notest_tags(self):
        doc = self.swagger_doc
        item_title = self.domain['people']['item_title']

        self.assertIn('tags', doc)
        tag_names = [tag['name'] for tag in doc['tags']]
        self.assertTrue(len([n for n in tag_names if n == item_title]) == 1)

        url = '/' + self.domain['people']['url']
        for m in ['get', 'post', 'delete']:
            path_m = doc['paths'][url][m]
            self.assertIn('tags', path_m)
            self.assertIn(item_title, path_m['tags'])

        url = '/%s/{%sId}' % (self.domain['people']['url'], item_title.lower())
        for m in ['get', 'patch', 'put', 'delete']:
            path_m = doc['paths'][url][m]
            self.assertIn('tags', path_m)
            self.assertIn(item_title, path_m['tags'])


if __name__ == '__main__':
    unittest.main()


eve_embedded/emdedded.py
"""
Módulo para embarcar dados de outras apis.

Usage:

No Settings:

'authors': {
    'type': 'list',
    'maxlength': 15,
    'schema': {"type": "string", "data_relation": {"api": "http://api.rbs.com.br/authors", "embeddable": True}}
},

No application.py:

from apis_common import embedded

application = Eve()
embedded.install(application)

Depende do módulo validator.py

Funcionamento:

Monkey patch que altera o método build_response_document para incluir
dados de outros serviços.

"""

from eve.methods import common
from eve.utils import parse_request, config, debug_error_message
import importlib
import json
import logging
import requests

default_headers = {"Accept": "text/json"}

# hook
before_build_document = []

logger = logging.getLogger(__name__)


def new_resolve_embedded_fields(resource, req, document):
    embedded_fields = []
    if req.embedded:
        try:
            client_embedding = json.loads(req.embedded)
        except ValueError:
            abort(400, description=debug_error_message(
                'Unable to parse `embedded` clause'
            ))
        try:
            embedded_fields = [k for k, v in client_embedding.items()
                               if v == 1]
        except AttributeError:
            # We got something other than a dict
            abort(400, description=debug_error_message(
                'Unable to parse `embedded` clause'
            ))

    embedded_fields = list(
        set(config.DOMAIN[resource]['embedded_fields']) |
        set(embedded_fields))
    enabled_embedded_fields = []
    for field in embedded_fields:
        field_def = new_field_definition(resource, field, document)
        if type(field_def) is dict:
            if field_def.get('type') == 'list':
                field_def = field_def['schema']
            if 'data_relation' in field_def and \
                    field_def['data_relation'].get('embeddable'):
                enabled_embedded_fields.append(field)

    return enabled_embedded_fields


def new_field_definition2(resource, chained_fields):

    definition = config.DOMAIN[resource]
    subfields = chained_fields.split('.')

    for field in subfields:
        if field not in definition.get('schema', {}):
            if 'data_relation' in definition and 'resource' in definition['data_relation']:
                sub_resource = definition['data_relation']['resource']
                definition = config.DOMAIN[sub_resource]
            else:
                definition = {'schema': {}}

        if field not in definition['schema']:
            return {}
        definition = definition['schema'][field]
        field_type = definition.get('type')
        if field_type == 'list':
            definition = definition['schema']
        elif field_type == 'objectid':
            pass
    return definition


def new_field_definition(resource, chained_fields, document):
    definition = config.DOMAIN[resource]
    subfields = chained_fields.split('.')

    for field in subfields:
        try:
            definition = definition['schema'][field]
            if definition.get('type') in ('list', 'string'):
                definition = definition['schema']
            elif definition['type'] == 'vars_meta':
                try:
                    tmpl = importlib.import_module(
                        'classifications_api.vars_meta_templates.{}'
                        .format(document['vars_meta_template']))
                    definition = {'schema': tmpl.schema}
                except ImportError:
                    pass

        except KeyError:
            return None

    return definition


def get_content(resource, content_id, additional_embedded={}):
    if not resource.startswith('http'):
        resource = 'http://api.rbs.com.br/' + resource
    parms = {}
    if additional_embedded:
        parms['embedded'] = json.dumps(dict((x, 1) for x in additional_embedded))
    try:
        embedded_req = requests.get('{}/{}'.format(resource, content_id), headers=default_headers, params=parms)
        embedded_doc = embedded_req.json()
        if embedded_req.status_code != 200:
            raise Exception("Error getting doc: {} - {}".format(embedded_req.status_code, str(embedded_doc)))
        if embedded_doc.get('_links'):
            del embedded_doc['_links']
    except:
        logger.error("Erro ao carregar conteúdo: {}/{}".format(resource, content_id), exc_info=True)
        embedded_doc = None
    return embedded_doc


def embedded_document(reference, data_relation, field_name, additional_embedded={}):
    return get_content(data_relation['api'], reference, additional_embedded)


def resolve_additional_embedded_documents(document, resource, embedded_fields):
    req = parse_request(resource)
    try:
        if req.embedded:
            extra_embedded = json.loads(req.embedded)
        else:
            extra_embedded = {}
    except ValueError:
        abort(400, description='Unable to parse `embedded` clause')
    extra_embedded = extra_embedded.keys()
    for field in new_resolve_embedded_fields(resource, req, document):
        field_extra_name = field + '.'
        field_extra_embedded = filter(lambda x: x.startswith(field_extra_name), extra_embedded)
        field_extra_embedded = map(lambda x: x.replace(field_extra_name, '', 1), field_extra_embedded)
        field_extra_embedded = list(field_extra_embedded)
        data_relation = new_field_definition(resource, field, document)['data_relation']
        getter = lambda ref: embedded_document(ref, data_relation, field, field_extra_embedded)
        fields_chain = field.split('.')
        last_field = fields_chain[-1]
        for subdocument in common.subdocuments(fields_chain[:-1], document=document, resource=resource):
            if last_field not in subdocument:
                continue
            if isinstance(subdocument[last_field], list):
                embedded_value = list(map(getter, subdocument[last_field]))
                embedded_value = [x for x in embedded_value if x is not None]
                subdocument[last_field] = embedded_value
            else:
                embedded_value = getter(subdocument[last_field])
                subdocument[last_field] = embedded_value


def install(app):
    original_build_response_document = common.build_response_document

    def new_build_response_document(document, resource, embedded_fields, latest_doc=None):
        for method in before_build_document:
            method(document, resource, embedded_fields, latest_doc)
        old_embedded_fields = filter(lambda x: new_field_definition(resource, x, None).get('data_relation').get('resource'), embedded_fields)
        old_embedded_fields = list(old_embedded_fields)
        new_embedded_fields = filter(lambda x: not new_field_definition(resource, x, None).get('data_relation').get('resource'), embedded_fields)
        new_embedded_fields = list(new_embedded_fields)
        result = original_build_response_document(document, resource, old_embedded_fields, latest_doc)
        resolve_additional_embedded_documents(document, resource, new_embedded_fields)
        return result

    common.field_definition = new_field_definition2
    common.build_response_document = new_build_response_document
    importlib.reload(importlib.import_module('eve.methods.get'))


eve_embedded/embedded.py
"""
Módulo para embarcar dados de outras apis.

Usage:

No Settings:

'authors': {
    'type': 'list',
    'maxlength': 15,
    'schema': {"type": "string", "data_relation": {"api": "http://api.com.br/authors", "embeddable": True}}
},

No application.py:

from eve_embedded import embedded

application = Eve()
embedded.install(application)

Funcionamento:

Monkey patch que altera o método build_response_document para incluir
dados de outros serviços.

"""

from eve.methods import common
from eve.utils import parse_request, config, debug_error_message
import importlib
import json
import logging
import requests

default_headers = {
    "Accept": "text/json",
    "Cache-Control": "no-cache",
}

# hook
before_build_document = []

extra_field_schemas = {}

logger = logging.getLogger(__name__)


def new_resolve_embedded_fields(resource, req, document):
    embedded_fields = []
    if req.embedded:
        try:
            client_embedding = json.loads(req.embedded)
        except ValueError:
            abort(400, description=debug_error_message(
                'Unable to parse `embedded` clause'
            ))
        try:
            embedded_fields = [k for k, v in client_embedding.items()
                               if v == 1]
        except AttributeError:
            # We got something other than a dict
            abort(400, description=debug_error_message(
                'Unable to parse `embedded` clause'
            ))

    embedded_fields = list(
        set(config.DOMAIN[resource]['embedded_fields']) |
        set(embedded_fields))
    enabled_embedded_fields = []
    for field in embedded_fields:
        field_def = new_field_definition(resource, field, document)
        if type(field_def) is dict:
            if field_def.get('type') == 'list':
                field_def = field_def['schema']
            if 'data_relation' in field_def and \
                    field_def['data_relation'].get('embeddable'):
                enabled_embedded_fields.append(field)

    return enabled_embedded_fields


def new_field_definition2(resource, chained_fields):

    definition = config.DOMAIN[resource]
    subfields = chained_fields.split('.')
    _keys = extra_field_schemas.keys()

    for field in subfields:
        field_type = definition.get('type')
        if field_type in _keys:
            definition = extra_field_schemas[field_type]
        if field not in definition.get('schema', {}):
            if 'data_relation' in definition and 'resource' in definition['data_relation']:
                sub_resource = definition['data_relation']['resource']
                definition = config.DOMAIN[sub_resource]
            else:
                definition = {'schema': {}}

        if field not in definition['schema']:
            return {}
        definition = definition['schema'][field]
        field_type = definition.get('type')
        if field_type in _keys:
            definition = extra_field_schemas[field_type]
        elif field_type == 'list':
            definition = definition['schema']
        elif field_type == 'objectid':
            pass
    return definition


def new_field_definition(resource, chained_fields, document):
    definition = config.DOMAIN[resource]
    subfields = chained_fields.split('.')

    _keys = extra_field_schemas.keys()
    for field in subfields:
        try:
            field_type = definition.get('type')
            if field_type in _keys:
                definition = extra_field_schemas[field_type]

            definition = definition['schema'][field]
            if definition.get('type') in ('list', 'string'):
                definition = definition['schema']
            elif definition['type'] == 'vars_meta':
                try:
                    tmpl = importlib.import_module(
                        'classifications_api.vars_meta_templates.{}'
                        .format(document['vars_meta_template']))
                    definition = {'schema': tmpl.schema}
                except ImportError:
                    pass

        except KeyError:
            return None

    return definition


def get_content(resource, content_id, additional_embedded={}):
    if not resource.startswith('http'):
        raise Exception('COULD NOT EMBED CONTENT BECAUSE RESOURCE DOES NOT START WITH HTTP. PLEASE CHECK DATA RELATION ON CLIENT')
    parms = {}
    if additional_embedded:
        parms['embedded'] = json.dumps(dict((x, 1) for x in additional_embedded))
    try:
        embedded_req = requests.get('{}/{}'.format(resource, content_id), headers=default_headers, params=parms)
        embedded_doc = embedded_req.json()
        if embedded_req.status_code != 200:
            raise Exception("Error getting doc: {} - {}".format(embedded_req.status_code, str(embedded_doc)))
        if embedded_doc.get('_links'):
            del embedded_doc['_links']
    except:
        logger.error("Error when load content: {}/{}".format(resource, content_id), exc_info=True)
        embedded_doc = None
    return embedded_doc


def embedded_document(reference, data_relation, field_name, additional_embedded={}):
    if type(reference) is str:
        return get_content(data_relation['api'], reference, additional_embedded)
    return reference


def resolve_additional_embedded_documents(document, resource, embedded_fields):
    req = parse_request(resource)
    try:
        if req.embedded:
            extra_embedded = json.loads(req.embedded)
        else:
            extra_embedded = {}
    except ValueError:
        abort(400, description='Unable to parse `embedded` clause')
    extra_embedded = extra_embedded.keys()
    for field in new_resolve_embedded_fields(resource, req, document):
        field_extra_name = field + '.'
        field_extra_embedded = filter(lambda x: x.startswith(field_extra_name), extra_embedded)
        field_extra_embedded = map(lambda x: x.replace(field_extra_name, '', 1), field_extra_embedded)
        field_extra_embedded = list(field_extra_embedded)
        data_relation = new_field_definition(resource, field, document)['data_relation']
        getter = lambda ref: embedded_document(ref, data_relation, field, field_extra_embedded)
        fields_chain = field.split('.')
        last_field = fields_chain[-1]
        for subdocument in common.subdocuments(fields_chain[:-1], document=document, resource=resource):
            if last_field not in subdocument:
                continue
            if isinstance(subdocument[last_field], list):
                embedded_value = list(map(getter, subdocument[last_field]))
                embedded_value = [x for x in embedded_value if x is not None]
                subdocument[last_field] = embedded_value
            else:
                embedded_value = getter(subdocument[last_field])
                subdocument[last_field] = embedded_value


def install(app, special_field_types=dict(type="special", schema={})):
    global extra_field_schemas
    extra_field_schemas = special_field_types
    original_build_response_document = common.build_response_document

    def new_build_response_document(document, resource, embedded_fields, latest_doc=None):
        for method in before_build_document:
            method(document, resource, embedded_fields, latest_doc)
        old_embedded_fields = filter(lambda x: new_field_definition(resource, x, None).get('data_relation').get('resource'), embedded_fields)
        old_embedded_fields = list(old_embedded_fields)
        new_embedded_fields = filter(lambda x: not new_field_definition(resource, x, None).get('data_relation').get('resource'), embedded_fields)
        new_embedded_fields = list(new_embedded_fields)
        result = original_build_response_document(document, resource, old_embedded_fields, latest_doc)
        resolve_additional_embedded_documents(document, resource, new_embedded_fields)
        return result

    common.field_definition = new_field_definition2
    common.build_response_document = new_build_response_document
    importlib.reload(importlib.import_module('eve.methods.get'))
